doc/crossfire-config.man
.TH crossfire-config 6 "Feb 04 2004" "" "CrossFire Game Server Manual"

Little program aimed at giving information to plugin about config of the
crossfire server.

Simply invoke with the config parameter to get. Only the most common parameters
(those that could be needed by an independent configure script) are available.
The rest is available in config.h andd should be included in any plugin needing
it.

    printf ("usage: crossfire-config --Dflags");
    printf (" (gives complete Dflags line for compiler invocation)\n");
    printf ("usage: crossfire-config --parameter-list");
    printf (" (show the list of available parameters)\n");
    printf ("usage: crossfire-config <parameter name>");
    printf (" (extract a compilation parameter)\n");


CHANGES
Changes in top level CVS tree.  A few notes when putting entries in this
file:
1) Please include a date in the form mm/dd/yyyy (yes, us americans are 
screwed up relative to most the world, but some consistent date format
is needed for this file.)
2) You don't need to go into exacting detail of the changes.  This file
is mostly meant to give an idea to others what has changed - you don't
need to recreate the code here.  If you change fixes a reported bug
on the mailing list, please provide a brief description of the bug
fixed.  But please do try to include a useful description of
what changed.  Saying something like 'General cleanup - just whitespace
changes' is much more useful than 'misc edits' or 'code cleanup'.
3) Please include your name with the changes you make.
4) Whenever making an update, please update this file, no matter how
small - this just lets others know that the file has changed if nothing
else.  With this, include the file(s) that you changed.
------------------------------------------------------------------------------

Changes for CVS:

This checkin mainly fixes multipart teleporters - they should now
work properly.  Also a minor fix to make the new pickup code work.
server/apply.c: apply_shop_mat rewritten to be more 'presentable'.
    Minor changes were needed in any case to pass the player/creature
    a a second object to teleport.
server/c_object.c: change the sscanf to be unsigned when getting value
    for new pickup mode.  At least on my system, if the value in the
    string was larger than the max signed value, sscanf set it to the 
    max value.
server/move.c: teleport function largely re-written.  Instead of the
    third arg being the 'originator', it is now the object to teleport.
    This allows the head of the teleporter to get passed to this function
    even if the object to teleport is not on the head.
server/time.c: move_teleporter modified to also check for objects on the
    other parts of multipart teleporters.  Also, pass the head object
    to the teleport function so that proper destination is used.
MSW 2001-09-02

Add Nils Lohner patch that new selections to the pickup mode.

include/define.h: Pickup type defines added.
include/player.h: Change mode (pickup mode) to 32 bits.
server/c_object.c: command_pickup function modified to support all
    the new pickup options.
server/player.c: Fix for path_to_player to prevent crashes.  Add check
    in move_player_attack just in case.  check_pick function modified to
    support new pickup options.  do_some_living modified to eat flesh
    items if player is starving and they don't have any food.
server/spell_effect.c: cast_detection modified so that detect magic/curse
    will not show objects beneath the floor.  Change so that detect_monster
    and detect_evil spells copy the face of what they detected - gives better
    appearance on client.
socket/item.c: fix bug LookAt that ws checking blocked_los on wrong
    offset.
MSW 2001-08-28

Enhancements to make monsters 'do the right thing' with relation
to tiled maps (ie, will detect players and attack them if on an adjoining
tiled maps).  This involved a rewrite of path_to_player - I've tested this
out, and if anything, it seems that monsters may be more aggressive on
spellcasting.
++
common/map.c: blocked modify to make sure it is checking valid coordinate
   value.  blocked_link,blocked_two modified to translate for tiled maps.
   Add load_and_link_tiled_map which loads a tiled map and links in all
   pointers relative to adjoining maps.  out_of_map modified to use
   this function instead of doing the work itself.  Add get_rangevector
   function that returns offset, direction, and distance between two
   objects, taking into account tiled maps.  get_rangevector_from_mapcoord
   added which is similar to get_rangevector, but takes a map and coordinates
   instead of a second object.  on_same_map function added which returns
   true if the two objects are on the same logical (ie tiled) maps.
doc/crossfire.doc: Add note about what Int does for monsters.
include/map.h: Add some comments, and add structure for rangevector
   function return values.
server/attack.c: add calls to on_same_map instead of doing op->map
   comparisons.
server/monster.c: Many functions modified to properly handle tiled
   maps properly.
server/pets.c: get_pet_enemy modified to know about tiled maps.
server/player.c: get_nearest_player modified to know about tiled maps.
   path_to_player rewritten to know about tiled maps - I think some other 
   bugs were also fixed in return values.  Split player_attack_door
   from move_player attack to keep functions a more reasonable size.
   move_player_attack, player_can_view modified to know about tiled
   maps.
server/skills.c: can detect_enemy modified to take another parameter -
   call in steal updated to pass this parameter.
server/spell_util.c: ok_to_put_more, can_see_monsterP modified to
   handle tiled maps properly.
socket/item.c: look_at modified to return proper results when player
   looks at something on an adjoining tiled map.
++ End of checkin

common/object.c:decrease_ob_nr modified to correctly update client when
    player uses item from container that is on the ground (bookshelf for
    example)
server/time.c: move_teleporter modified to skip over wall/floors and
    teleport whatever is above them.
MSW 2001-08-05

common/map.c: modify link_multipart_objects to preserve name and title of
    head when making the more parts.  modify update_position so that it should
    never show an invisible face to the player.
common/object.c: fix check_walk_on so that spells will now properly get
    affected by spinners and directories.
include/sockproto.h: Rebuilt to include MapNewmapCmd
server/hiscore.c: Fix crash if player savebeds map on map which does not
    have a name.  Use the path instead.
server/move.c: modify teleporter function to send MapNewmapCmd to client

devel/crossfire-config.c
/*
    CrossFire, A Multiplayer game for X-windows

    Copyright (C) 2001 Mark Wedel & Crossfire Development Team
    Copyright (C) 1992 Frank Tore Johansen

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    The authors can be reached via e-mail to crossfire-devel@real-time.com
*/

/*
 * Small program that gives plugins information about the compile-time
 * configuration of the server. Only the most common parameters (those that
 * might be needed by an independent configure script) are available. The rest
 * can be found in 'config.h' and should be included in any plugin needing it.
 */

#include "stdio.h"
#include "stdlib.h"
#include "string.h"

typedef struct {
    const char *name;
    const char *value;
} cf_parameter;

const cf_parameter cf_parameter_list[] = {
    { "CONFDIR", CONFDIR },
    { "DATADIR", DATADIR },
    { "LIBDIR", LIBDIR },
    { "LOCALDIR", LOCALDIR },
    { "PLUGIN_SUFFIX", PLUGIN_SUFFIX },
};

const int cf_parameter_list_size =
    sizeof(cf_parameter_list) / sizeof(cf_parameter);

static void print_usage() {
    fprintf(stderr,
        "usage: crossfire-config <parameter>\n"
        "       crossfire-config [options]\n"
        "\n"
        "Options:\n"
        "    --Dflags            show complete Dflags line from compiler invocation\n"
        "    --parameter-list    show the list of available parameters\n"
        "\n"
    );
}

int main(int argc, char *argv[]) {
    int i;

    // Print usage instructions when given invalid command-line arguments.
    if (argc != 2) {
        print_usage();
        exit(EXIT_FAILURE);
    }

    if (strcmp(argv[1], "--Dflags") == 0) {
        for (i = 0; i < cf_parameter_list_size; i++) {
            printf("-D%s=\\\"%s\\\" ", cf_parameter_list[i].name,
                    cf_parameter_list[i].value);
        }

        printf("\n");
        exit(EXIT_SUCCESS);
    }

    // Show a list of available parameters.
    if (strcmp(argv[1], "--parameter-list") == 0) {
        printf("Available parameters:\n");

        // PLUGININSTALLDIR is a special parameter handled separately.
        printf("    PLUGININSTALLDIR\n");

        for (i = 0; i < cf_parameter_list_size; i++) {
            printf("    %s\n", cf_parameter_list[i].name);
        }

        exit(EXIT_SUCCESS);
    }

    // Handle plugin installation directory as a special case.
    if (strcmp(argv[1], "PLUGININSTALLDIR") == 0) {
        printf("%s/plugins/\n", LIBDIR);
        exit(EXIT_SUCCESS);
    }

    for (i = 0; i < cf_parameter_list_size; i++) {
        if (!strcmp(argv[1], cf_parameter_list[i].name)) {
            printf("%s\n", cf_parameter_list[i].value);
            exit(EXIT_SUCCESS);
        }
    }

    exit(EXIT_FAILURE);
}
